lib/Command/Generate.php

```diff
@@ -142,38 +148,77 @@ private function generatePathPreviews(IUser $user, string $path) {
 			return;
 		}
 		$pathFolder = $userFolder->get($relativePath);
-		$this->parseFolder($pathFolder);
+		$this->processFolder($pathFolder, $user);
 	}
 
 	private function generateUserPreviews(IUser $user) {
 		\OC_Util::tearDownFS();
 		\OC_Util::setupFS($user->getUID());
 
 		$userFolder = $this->rootFolder->getUserFolder($user->getUID());
-		$this->parseFolder($userFolder);
+		$this->processFolder($userFolder, $user);
 	}
 
-	private function parseFolder(Folder $folder) {
+	private function processFolder(Folder $folder, IUser $user) {
 		// Respect the '.nomedia' file. If present don't traverse the folder
 		if ($folder->nodeExists('.nomedia')) {
 			$this->output->writeln('Skipping folder ' . $folder->getPath());
 			return;
 		}
 
+		// random sleep between 0 and 50ms to avoid collision between 2 processes
+		usleep(rand(0,50000));
```

Comment:
Could we not get around this with a DB transaction?

Reply:
Indeed, you could get a lock by querying a row where file is node.id and locked is false, setting locked to true.
If this returns a result, you have the lock. If no lock, skip/wait.

Reply:
Hi,

I am not sure I understand what you both mean. Ideally we wouldn't have to do this if we could use some kind of `pthreads`/`parallel` approach but since ZST packages are not available in current Debian systems, or even in Sury, I am calling the `occ` command several times in parallel.

> Indeed, you could get a lock by querying a row where file is node.id and locked is false, setting locked to true.
If this returns a result, you have the lock. If no lock, skip/wait.

Since we can't control different threads, but we have independent processes, then I added the `lock` column to the database so that two different instances don't start working on the same preview at the same time.

If you look at the code, that is exactly what I do, but because there is a chance that both instances read the lock value as unlocked at the same time before setting it as locked, I added an ugly random wait to minimize the chance of this happening.

This wouldn't be needed if we had multithread approach with barriers, waits and so on. (or just give each child thread a different group of previews to work on).

I hope it is clear now. This is not particularly pretty but given the limitations we have it has proven very much effective :)

Reply:
I am no PHP or SQL coder, but isn't there a way to combine a lock check + lock to eliminate the race condition?

E.g. in shell I do something like this:
```
set -C
if { > process.lock; } 2> /dev/null; then
  start_process
else
  skip_process
fi
set +C
```
- `set -C` adds the noclobber option, thus nothing can be written to existing files.
- Assuming that two threads cannot write the same file at the same time, one write (empty redirect to lock file) must fail, which results in process is skipped.
- So basically checking for a lock and creating the lock is one and the same quick task.
- Isn't there something similar possible for the preview generation processes, through PHP or SQL?

Reply:
The idea is to ran several instances simultaneously

```
for i in `seq 1 16`; do
  occ_command &
done
wait
```

I want several processes in parallel, so the locking needs to be in the PHP section

Reply:
Currently I see a race condition for multiple processes started at the same time.
```
get record
if record exists:
   if not locked:
      lock record (update)
      process record
      finally delete record --> this could make it get picked up again by another process
   else:
      pass
else:
    create (insert) locked record
    process record
    finally delete record --> this could make it get picked up again by another process
```
It could be as simple as
```
foreach file
   create record (unique filename, pid)
   if create successful:
      process record
   else:
      pass, as another process created the record

# after all files have been processed
delete each record matching pid
# to ensure they get processed again on future runs
```
The records are locked here until the process finishes.

Reply:
If you start 16 processes at the same time with the snippet above, I'd expect them to go like this:
```
pid1: img1, img17, img33, ...
pid2: img2, img18, img34, ...
...
```
If you start 8 processes at the same time and then 8 half an hour later, the second batch will quickly catch up with the first batch, spread the load over 16 pids and they finish around the same time.

If you process all images, then wait 10 minutes, then start another run, every file will get processed again (as intended, right?).

Reply:
Hi,

Thanks for joining the discussion. Yes, I agree that delaying the deletion until the very end would be more robust. In that case we would also replace the `finally` with a `catch`.

We are thinking in a slightly different use case since I am starting all processes at the same(ish) time so I wasn't thinking about the use case you describe, but deleting at the very end would also cover that case so that would be a good improvement.

We are talking all the time about the `generate-all` command which would only be ran once in the case where we install the app after we already have uploaded files. Normally we run the `pre-generate` command which uses the database instead of traversing the filesystem and doesn't have this problem.

Still in both approaches we have this probability for a race condition which is why I added a small random delay, but I don't see a way around that other than using PHP multithreading to 1) list all files 2) split the file list between threads.

Reply:
edit: for sure there will be improvements to be made but before spending too much time on the small print I would like to see what @rullzer thinks about the whole idea to see if it's worth to keep spending time on this or this is not going to be accepted in any form.

Reply:
@nachoparker 
Still on the race condition topic:
- Currently a table row `locked` is created with values `true` or `false` being set.
- I am not sure what is possible with this Nextcloud -> PHP -> SQL API, but is there a way/function which assigns values to empty fields only?
- In case it as well is possible to check some exit code of the field assignment, this could be used as condition, else the PID could be added as value for the `locked` fields and checked back if matching the current process.
- But all this requires a function to create/ write to fields only if non-existent/empty, to eliminate the race condition between check first and assign afterwards.

Reply:
Don't need return code which checks for success (you are thinking shell scripting), we can just work with SQL.

There's many ways to skin a cat. Deleting at the very end of the process would be an easy improvement but I don't want to keep going if this is going nowhere so I am awaiting feedback from the maintainer.

Reply:
I've been working on a python version of the preview generator with [jpegtran](https://pypi.org/project/jpegtran-cffi/) for blazing fast image processing.

This snippet allows running multiple python processes.
```python
for fileid, storage, name, path in image_filecache_query:                             
    # check preview dir                                              
    try:                                                             
        cursor.execute("INSERT INTO sg_preview (path, pid) VALUES ('{0}_{1}', {2})".format(storage, path, os.getpid()))
        if not cursor:                                               
            continue                                                 
        conn.commit()                                                
    except MySQLdb.Error:                                            
        continue                                                     
    process_image(...)
```
I'll let you know when it works properly :)

Reply:
With Python as dependency IMO not a solution for the default app, but nice as alternative, if it's finally more efficient/faster and/or for systems where Python is available anyway.

Reply:
@stefaang sounds cool, great. I imagine it spawns threads in parallel or uses multithreading per image?

That would be a nice thing to have even though ideally this would be implemented in multithreading PHP in the long run.

But hey, that might never happen so kudos

Reply:
@nachoparker
> Don't need return code which checks for success (you are thinking shell scripting), we can just work with SQL to see if the column is locked or a new PID column matches, that's quite straightforward.

That would not solve the race condition. The problem is the time between checking the fields state/value and setting new state/value. E.g. two parallel processes/threads by chance check the same field nearly the same time. One checks slightly earlier but it can take slightly too long for it to set the locked state/value, so the other process still finds the filed unlocked. So generally when doing a check and setting a value based on the checks result, this strictly inherits such a race condition. For this reason it would be great to skip the check and set the value directly. But this requires a set function that fails conditionally and this failure must be recognised afterwards to decide whether to go on processing or skipping it.